File: examples/billboard/billboard.py
# -*- Mode:Python -*-

##########################################################################
#                                                                        #
# This file is part of AVANGO.                                           #
#                                                                        #
# Copyright 1997 - 2009 Fraunhofer-Gesellschaft zur Foerderung der       #
# angewandten Forschung (FhG), Munich, Germany.                          #
#                                                                        #
# AVANGO is free software: you can redistribute it and/or modify         #
# it under the terms of the GNU Lesser General Public License as         #
# published by the Free Software Foundation, version 3.                  #
#                                                                        #
# AVANGO is distributed in the hope that it will be useful,              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of         #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           #
# GNU General Public License for more details.                           #
#                                                                        #
# You should have received a copy of the GNU Lesser General Public       #
# License along with AVANGO. If not, see <http://www.gnu.org/licenses/>. #
#                                                                        #
##########################################################################

import avango.osg
import avango.display
import sys

class Rotation(avango.script.Script):
    TimeIn = avango.SFFloat()
    MatrixOut = avango.osg.SFMatrix()

    def evaluate(self):
        self.MatrixOut.value = avango.osg.make_rot_mat(self.TimeIn.value, 0, 1, 0)


argv = avango.display.init(sys.argv)

text1 = avango.osg.nodes.Text(String = "LayerGeode",
                            Size = 0.2,
                            Alignment = 4,
                            Color = avango.osg.Vec4(0,0,0,1)
                            )
text2 = avango.osg.nodes.Text(String = "Billboard",
                            Size = 0.2,
                            Alignment = 4,
                            Color = avango.osg.Vec4(0,0,0,1)
                            )
panel1 = avango.osg.nodes.Panel(PanelColor = avango.osg.Vec4(1,1,0,0.5),
                            Width = 0.7,
                            Height = 0.25,
                            )
panel2 = avango.osg.nodes.Panel(PanelColor = avango.osg.Vec4(1,0,0,0.5),
                            Width = 0.7,
                            Height = 0.25,
                            )

layer_geode = avango.osg.nodes.LayerGeode(Drawables = [panel1, text1])

billboard = avango.osg.nodes.Billboard(Drawables = [panel2, text2])
billboard.Mode.value = 2  # AXIAL_ROT
billboard.Axis.value = avango.osg.Vec3(0.,1.,0.)
billboard.Normal.value = avango.osg.Vec3(0.0,0.0,1.0)

translation_layer_geode = avango.osg.nodes.MatrixTransform(
    Children=[layer_geode],
    Matrix =  avango.osg.make_trans_mat(0.8,0.2,0))
translation_billboard = avango.osg.nodes.MatrixTransform(
    Children=[billboard],
    Matrix =  avango.osg.make_trans_mat(0.8,-0.2,0))
rotation_boards = avango.osg.nodes.MatrixTransform(
    Children=[translation_billboard, translation_layer_geode])
translation_world = avango.osg.nodes.MatrixTransform(
    Children=[rotation_boards],
    Matrix = avango.osg.make_trans_mat(0,1.7,-13))

time_sensor = avango.nodes.TimeSensor()

rotation = Rotation()
rotation_boards.Matrix.connect_from(rotation.MatrixOut)
rotation.TimeIn.connect_from(time_sensor.Time)

root = avango.osg.nodes.Group(Name="Root")
root.Children.value = [ translation_world ]
view = avango.display.make_view()
view.Root.value = root

avango.display.run()

File: attic/avango-osg/src/avango/osg/Billboard.cpp
// -*- Mode:C++ -*-

/************************************************************************\
*                                                                        *
* This file is part of Avango.                                           *
*                                                                        *
* Copyright 1997 - 2008 Fraunhofer-Gesellschaft zur Foerderung der       *
* angewandten Forschung (FhG), Munich, Germany.                          *
*                                                                        *
* Avango is free software: you can redistribute it and/or modify         *
* it under the terms of the GNU Lesser General Public License as         *
* published by the Free Software Foundation, version 3.                  *
*                                                                        *
* Avango is distributed in the hope that it will be useful,              *
* but WITHOUT ANY WARRANTY; without even the implied warranty of         *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           *
* GNU General Public License for more details.                           *
*                                                                        *
* You should have received a copy of the GNU Lesser General Public       *
* License along with Avango. If not, see <http://www.gnu.org/licenses/>. *
*                                                                        *
* Avango is a trademark owned by FhG.                                    *
*                                                                        *
\************************************************************************/

#include <boost/bind.hpp>
#include <avango/osg/Billboard.h>
#include <avango/Logger.h>
#include <osg/Version>

namespace
{
av::Logger& logger(av::getLogger("av::osg::Billboard"));
}

AV_FC_DEFINE(av::osg::Billboard);

AV_FIELD_DEFINE(av::osg::SFBillboard);
AV_FIELD_DEFINE(av::osg::MFBillboard);

av::osg::Billboard::Billboard(::osg::Billboard* osgbillboard) : Geode(osgbillboard), mOsgBillboard(osgbillboard)
{
    AV_FC_ADD_ADAPTOR_FIELD(Mode, boost::bind(&Billboard::getModeCB, this, _1), boost::bind(&Billboard::setModeCB, this, _1));
    AV_FC_ADD_ADAPTOR_FIELD(Axis, boost::bind(&Billboard::getAxisCB, this, _1), boost::bind(&Billboard::setAxisCB, this, _1));
    AV_FC_ADD_ADAPTOR_FIELD(Normal, boost::bind(&Billboard::getNormalCB, this, _1), boost::bind(&Billboard::setNormalCB, this, _1));
}

av::osg::Billboard::~Billboard() {}

void av::osg::Billboard::initClass()
{
    if(!isTypeInitialized())
    {
        av::osg::Geode::initClass();

        AV_FC_INIT(av::osg::Geode, av::osg::Billboard, true);

        SFBillboard::initClass("av::osg::SFBillboard", "av::Field");
        MFBillboard::initClass("av::osg::MFBillboard", "av::Field");
    }
}

osg::Billboard* av::osg::Billboard::getOsgBillboard() const { return mOsgBillboard; }

/* virtual */ void av::osg::Billboard::getModeCB(const SFUInt::GetValueEvent& event) { *(event.getValuePtr()) = mOsgBillboard->getMode(); }

/* virtual */ void av::osg::Billboard::setModeCB(const SFUInt::SetValueEvent& event) { mOsgBillboard->setMode(static_cast<::osg::Billboard::Mode>(event.getValue())); }

/* virtual */ void av::osg::Billboard::getAxisCB(const av::osg::SFVec3::GetValueEvent& event) { *(event.getValuePtr()) = mOsgBillboard->getAxis(); }

/* virtual */ void av::osg::Billboard::setAxisCB(const av::osg::SFVec3::SetValueEvent& event) { mOsgBillboard->setAxis(event.getValue()); }

/* virtual */ void av::osg::Billboard::getNormalCB(const av::osg::SFVec3::GetValueEvent& event) { *(event.getValuePtr()) = mOsgBillboard->getNormal(); }

/* virtual */ void av::osg::Billboard::setNormalCB(const av::osg::SFVec3::SetValueEvent& event) { mOsgBillboard->setNormal(event.getValue()); }
